aiohttp/multipart.py

```diff
@@ -938,53 +682,191 @@ def __exit__(self, exc_type, exc_val, exc_tb):
         pass
 
     def __iter__(self):
-        return iter(self.parts)
+        return iter(self._parts)
 
     def __len__(self):
-        return len(self.parts)
+        return len(self._parts)
 
     @property
     def boundary(self):
-        *_, params = parse_mimetype(self.headers.get(CONTENT_TYPE))
-        return params['boundary'].encode('us-ascii')
+        return self._boundary
 
     def append(self, obj, headers=None):
-        """Adds a new body part to multipart writer."""
-        if isinstance(obj, self.part_writer_cls):
-            if headers:
+        if headers is None:
+            headers = CIMultiDict()
+
+        if isinstance(obj, Payload):
+            if obj.headers is not None:
                 obj.headers.update(headers)
-            self.parts.append(obj)
+            else:
+                obj._headers = headers
+            self.append_payload(obj)
         else:
-            if not headers:
-                headers = CIMultiDict()
-            self.parts.append(self.part_writer_cls(obj, headers))
-        return self.parts[-1]
+            try:
+                self.append_payload(get_payload(obj, headers=headers))
+            except LookupError:
+                raise TypeError
+
+    def append_payload(self, payload):
+        """Adds a new body part to multipart writer."""
+        # content-type
+        if CONTENT_TYPE not in payload.headers:
+            payload.headers[CONTENT_TYPE] = payload.content_type
+
+        # compression
+        encoding = payload.headers.get(CONTENT_ENCODING, '').lower()
+        if encoding and encoding not in ('deflate', 'gzip', 'identity'):
+            raise RuntimeError('unknown content encoding: {}'.format(encoding))
+        if encoding == 'identity':
+            encoding = None
+
+        # te encoding
+        te_encoding = payload.headers.get(
+            CONTENT_TRANSFER_ENCODING, '').lower()
+        if te_encoding not in ('', 'base64', 'quoted-printable', 'binary'):
+            raise RuntimeError('unknown content transfer encoding: {}'
+                               ''.format(te_encoding))
+        if te_encoding == 'binary':
+            te_encoding = None
+
+        # size
+        size = payload.size
+        if size is not None and not (encoding or te_encoding):
+            payload.headers[CONTENT_LENGTH] = str(size)
+
+        # render headers
+        headers = ''.join(
+            [k + ': ' + v + '\r\n' for k, v in payload.headers.items()]
+        ).encode('utf-8') + b'\r\n'
+
+        self._parts.append((payload, headers, encoding, te_encoding))
 
     def append_json(self, obj, headers=None):
         """Helper to append JSON part."""
-        if not headers:
+        if headers is None:
             headers = CIMultiDict()
-        headers[CONTENT_TYPE] = 'application/json'
-        return self.append(obj, headers)
+
+        *_, params = parse_mimetype(headers.get(CONTENT_TYPE))
+        charset = params.get('charset', 'utf-8')
```

Comment:
I think this is a bit redundant and may confuse. Because you can throw here any content-type, but in the end it will be `application/json` without any charset and else params. 

Better have explicit charset function argument or drop this idea. We're in 2017 anyway and Python`s json escapes all non ascii chars, so it doesn't really matter which encoding will be applied.